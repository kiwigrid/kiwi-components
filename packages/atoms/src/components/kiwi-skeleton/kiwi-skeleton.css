.loading-skeleton {
  --line-height: 18px;
  --fill: #bdbcbc;
  --fill-highlight: #e2e2e2;
}

.loading-skeleton li {
  margin-top: 8px;
}

@keyframes skeleton-loading {
  0% {
    background-position: 0% 50%;
  }
  50% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0% 50%;
  }
}

/* Paragraph */

.loading-skeleton.paragraph li,
.loading-skeleton.paragraph h3 {
  height: var(--line-height);
  background: linear-gradient(
    90deg,
    var(--fill-highlight),
    var(--fill) 50%,
    var(--fill-highlight)
  );
  background-size: 400% 100%;
  animation: skeleton-loading 1.8s infinite;
  border-radius: 2px;
}

.loading-skeleton.paragraph h3 {
  width: 50%;
  margin: var(--line-height) 0;
}

/* Table */
.loading-skeleton.table {
  --line-height: 32px;
}

.loading-skeleton.table h3 {
  margin: var(--line-height) 0;
}

.loading-skeleton.table h3,
.loading-skeleton.table li {
  display: flex;
}

.loading-skeleton.table span {
  display: block;
  width: 100%;
  margin: 0 8px;
  height: var(--line-height);
  background: linear-gradient(
    90deg,
    var(--fill-highlight),
    var(--fill) 50%,
    var(--fill-highlight)
  );
  background-size: 400% 100%;
  animation: skeleton-loading 1.8s infinite;
  border-radius: 2px;
}

.loading-skeleton.table span:first-child {
  width: 33%;
}

.loading-skeleton.table span:last-child {
  width: 66%;
}
