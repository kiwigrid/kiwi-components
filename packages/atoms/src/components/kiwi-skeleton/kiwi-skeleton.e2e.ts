import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-skeleton', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-skeleton></kiwi-skeleton>');

    const element = await page.find('kiwi-skeleton');
    expect(element).toHaveClass('hydrated');
  });
});
