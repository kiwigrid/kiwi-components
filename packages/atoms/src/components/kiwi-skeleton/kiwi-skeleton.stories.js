import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Skeleton',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const paragraph = (args) => {
  const { rows } = args;

  return html`
    <div class="panel panel-default">
      <div class="panel-body">
        <kiwi-skeleton type="paragraph" rows=${rows}>
      </div>
    </div>
  `;
};

export const table = (args) => {
  const { rows } = args;

  return html`
    <div class="panel panel-default">
      <div class="panel-body">
        <kiwi-skeleton type="table" rows=${rows}>
      </div>
    </div>
  `;
};

paragraph.args = table.args = {
  rows: 4,
};
