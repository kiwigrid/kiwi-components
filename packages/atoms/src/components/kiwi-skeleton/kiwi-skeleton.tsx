import {
  Component,
  ComponentInterface,
  h,
  JSX,
  Prop,
  FunctionalComponent,
} from '@stencil/core';
import { range } from 'lodash-es';

@Component({
  tag: 'kiwi-skeleton',
  styleUrl: 'kiwi-skeleton.css',
  shadow: false,
})
export class KiwiSkeleton implements ComponentInterface {
  /**
   * Number of skeleton rows.
   */
  @Prop()
  rows = 4;

  /**
   * The kind of element this skeleton should represent.
   */
  @Prop()
  type: 'paragraph' | 'table' = 'paragraph';

  render(): JSX.Element {
    switch (this.type) {
      case 'paragraph':
        return <Paragraph rows={this.rows} />;
      case 'table':
        return <Table rows={this.rows} />;
    }
  }
}

const Paragraph: FunctionalComponent<{ rows: number }> = ({ rows }) => (
  <div class="loading-skeleton paragraph">
    <h3></h3>
    <ul class="list-unstyled">
      {range(0, rows).map(() => (
        <li></li>
      ))}
    </ul>
  </div>
);

const Table: FunctionalComponent<{ rows: number }> = ({ rows }) => (
  <div class="loading-skeleton table">
    <h3>
      <span></span>
      <span></span>
      <span></span>
      <span></span>
    </h3>
    <ul class="list-unstyled">
      {range(0, rows).map(() => (
        <li>
          <span></span>
          <span></span>
          <span></span>
          <span></span>
        </li>
      ))}
    </ul>
  </div>
);
