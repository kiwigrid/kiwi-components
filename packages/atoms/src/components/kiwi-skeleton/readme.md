# kiwi-skeleton



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                         | Type                     | Default       |
| -------- | --------- | --------------------------------------------------- | ------------------------ | ------------- |
| `rows`   | `rows`    | Number of skeleton rows.                            | `number`                 | `4`           |
| `type`   | `type`    | The kind of element this skeleton should represent. | `"paragraph" \| "table"` | `'paragraph'` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
