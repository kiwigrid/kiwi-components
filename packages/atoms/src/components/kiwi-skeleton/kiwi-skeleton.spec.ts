import { newSpecPage } from '@stencil/core/testing';
import { KiwiSkeleton } from './kiwi-skeleton';

describe('kiwi-skeleton', () => {
  it('builds', () => {
    expect(new KiwiSkeleton()).toBeTruthy();
  });

  describe('paragraph', () => {
    it('renders', async () => {
      const skeleton = await newSpecPage({
        components: [KiwiSkeleton],
        html: `
        <kiwi-skeleton />
      `,
      });

      expect(skeleton.root).toMatchSnapshot('default rows');
    });

    it('renders 6 rows', async () => {
      const skeleton = await newSpecPage({
        components: [KiwiSkeleton],
        html: `
        <kiwi-skeleton rows="6" />
      `,
      });

      expect(skeleton.root).toMatchSnapshot('6 rows');
    });
  });

  describe('table', () => {
    it('renders', async () => {
      const skeleton = await newSpecPage({
        components: [KiwiSkeleton],
        html: `
        <kiwi-skeleton type="table"/>
      `,
      });

      expect(skeleton.root).toMatchSnapshot();
    });
  });
});
