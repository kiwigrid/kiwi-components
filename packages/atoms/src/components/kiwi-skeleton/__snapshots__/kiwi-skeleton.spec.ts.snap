// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-skeleton paragraph renders 6 rows: 6 rows 1`] = `
<kiwi-skeleton rows="6">
  <div class="loading-skeleton paragraph">
    <h3></h3>
    <ul class="list-unstyled">
      <li></li>
      <li></li>
      <li></li>
      <li></li>
      <li></li>
      <li></li>
    </ul>
  </div>
</kiwi-skeleton>
`;

exports[`kiwi-skeleton paragraph renders: default rows 1`] = `
<kiwi-skeleton>
  <div class="loading-skeleton paragraph">
    <h3></h3>
    <ul class="list-unstyled">
      <li></li>
      <li></li>
      <li></li>
      <li></li>
    </ul>
  </div>
</kiwi-skeleton>
`;

exports[`kiwi-skeleton table renders 1`] = `
<kiwi-skeleton type="table">
  <div class="loading-skeleton table">
    <h3>
      <span></span>
      <span></span>
      <span></span>
      <span></span>
    </h3>
    <ul class="list-unstyled">
      <li>
        <span></span>
        <span></span>
        <span></span>
        <span></span>
      </li>
      <li>
        <span></span>
        <span></span>
        <span></span>
        <span></span>
      </li>
      <li>
        <span></span>
        <span></span>
        <span></span>
        <span></span>
      </li>
      <li>
        <span></span>
        <span></span>
        <span></span>
        <span></span>
      </li>
    </ul>
  </div>
</kiwi-skeleton>
`;
